### PR TITLE
UCT/MLX5: Preparations for HW TM for PRM tls p1

### PR DESCRIPTION
**Pull request by brminich**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5709/ for details (Mellanox internal link).

**Comment by brminich** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5712/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
hang on hpc-test-node:
```
12:47:32 Note: This is test shard 1 of 4.
12:47:32 Note: Randomizing tests' orders with a seed of 93984 .
12:47:32 [==========] Running 2263 tests from 520 test cases.
12:47:32 [----------] Global test environment set-up.
12:47:32 [----------] 9 tests from cmrcx/test_ucp_rma
12:47:32 [ RUN      ] cmrcx/test_ucp_rma.blocking_large/1
12:47:33 [     SKIP ] (skipping on valgrind)
12:47:33 [       OK ] cmrcx/test_ucp_rma.blocking_large/1 (1367 ms)
12:47:33 [ RUN      ] cmrcx/test_ucp_rma.nonblocking_stream_put_nbi_flush_worker/1
12:47:34 ==11035== WARNING: valgrind ignores shmget(shmflg) SHM_HUGETLB
12:47:36 [       OK ] cmrcx/test_ucp_rma.nonblocking_stream_put_nbi_flush_worker/1 (2263 ms)
12:47:36 [ RUN      ] cmrcx/test_ucp_rma.nbi_med/1
16:00:59 Build was aborted
```

**Comment by Di0gen** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5725/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5738/ for details (Mellanox internal link).

**Comment by brminich** [comment]
unrelated arm env failures

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5739/ for details (Mellanox internal link).

**Comment by brminich** [comment]
@yosefe, plz check

File: src/uct/ib/rc/accel/rc_mlx5_common.h [hunk 1]
```diff
@@ -219,6 +180,84 @@ uct_rc_mlx5_iface_common_poll_rx(uct_rc_mlx5_iface_common_t *mlx5_common_iface,
...
+    }
+
+    ucs_memory_cpu_load_fence();
+    UCS_STATS_UPDATE_COUNTER(rc_iface->stats, UCT_RC_IFACE_STAT_RX_COMPLETION, 1);
+
+    byte_len = ntohl(cqe->byte_cnt);
+    seg      = uct_ib_mlx5_srq_get_wqe(&mlx5_common_iface->rx.srq,
+                                       ntohs(cqe->wqe_counter));
+
+    uct_rc_mlx5_iface_common_am_handler(mlx5_common_iface, rc_iface, cqe,
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5_common.h]: what is the reason for moving this to separate function?

File: src/uct/ib/rc/base/rc_iface.h [hunk 2]
```diff
@@ -361,13 +390,9 @@ uct_rc_iface_tag_imm_data_pack(uint32_t *ib_imm, uint32_t *app_ctx,
...
 static UCS_F_ALWAYS_INLINE uint64_t
-uct_rc_iface_tag_imm_data_unpack(struct ibv_exp_wc *wc, uint32_t app_ctx)
+uct_rc_iface_tag_imm_data_unpack(uint32_t ib_imm, uint32_t app_ctx, int is_imm)
 {
-    if (wc->exp_wc_flags & IBV_EXP_WC_WITH_IMM) {
-        return ((uint64_t)app_ctx << 32) | wc->imm_data;
-    } else {
-        return 0ul;
-    }
+    return is_imm ? ((uint64_t)app_ctx << 32) | ib_imm : 0ul;
```
Comment (yosefe) [inline src/uct/ib/rc/base/rc_iface.h]: maybe add more ( ) here around what is between ? and :

File: src/uct/ib/rc/accel/rc_mlx5_common.h (hunk 1 above)
Comment (brminich) [reply src/uct/ib/rc/accel/rc_mlx5_common.h]: Tag RX poll function will invoke it as well when NO_TAG message arrives

File: src/uct/ib/rc/base/rc_iface.h (hunk 2 above)
Comment (brminich) [reply src/uct/ib/rc/base/rc_iface.h]: ok

File: src/uct/ib/rc/accel/rc_mlx5_common.h (hunk 1 above)
Comment (yosefe) [reply src/uct/ib/rc/accel/rc_mlx5_common.h]: will there be separate CQ for tags?

Comment (brminich) [reply src/uct/ib/rc/accel/rc_mlx5_common.h]: CQ will be the same, but the progress function will be different, like with verbs

Comment (yosefe) [reply src/uct/ib/rc/accel/rc_mlx5_common.h]: seems like code duplication.. maybe we can have on progress function which have a "boolean" parameter, resolved in compile time, which tells if TM CQE types are to be checked?

**Comment by brminich** [comment]
@yosefe, rebased

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5759/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5760/ for details (Mellanox internal link).

**Pull request merged**